src/beamlime/applications/daemons.py

```diff
@@ -0,0 +1,173 @@
+        ] = dict()
+        self.message_pipe = list()
+        self._break_routing_loop = False  # Break the routing loop flag
+
+    def break_routing_loop(self, message: Optional[BeamlimeMessage] = None) -> None:
+        if not isinstance(message, self.StopRouting):
+            raise TypeError(
+                f"Expected message of type {self.StopRouting}, got {type(message)}."
+            )
+        self._break_routing_loop = True
+
+    @contextmanager
+    def _handler_wrapper(
+        self, handler: Callable[..., Any], message: BeamlimeMessage
+    ) -> Generator[None, None, None]:
+        import warnings
+
+        try:
+            self.debug(f"Routing event {type(message)} to handler {handler}...")
+            yield
+        except Exception as e:
+            warnings.warn(f"Failed to handle event {type(message)}: {e}", stacklevel=2)
+        else:
+            self.debug(f"Routing event {type(message)} to handler {handler} done.")
+
+    def _register(
+        self,
+        *,
+        handler_list: dict[
+            type[BeamlimeMessage], List[Callable[[BeamlimeMessage], Any]]
+        ],
+        event_tp: type[BeamlimeMessage],
+        handler: Callable[[BeamlimeMessage], Any],
+    ):
+        if event_tp in handler_list:
+            handler_list[event_tp].append(handler)
+        else:
+            handler_list[event_tp] = [handler]
+
+    def register_awaitable_handler(
+        self, event_tp, handler: Callable[[BeamlimeMessage], Coroutine[Any, Any, Any]]
+    ):
+        self._register(
+            handler_list=self.awaitable_handlers, event_tp=event_tp, handler=handler
+        )
+
+    def register_handler(self, event_tp, handler: Callable[[BeamlimeMessage], Any]):
+        self._register(handler_list=self.handlers, event_tp=event_tp, handler=handler)
+
+    def _check_result(self, result: Any) -> None:
+        """Append or extend ``result`` to ``self.message_pipe``.
+
+        It filters out non-BeamlimeMessage objects from ``result``.
+        """
+        if isinstance(result, BeamlimeMessage):
+            self.message_pipe.append(result)
+        elif isinstance(result, tuple):
+            msgs = (_msg for _msg in result if isinstance(_msg, BeamlimeMessage))
+            self.message_pipe.extend(msgs)
+
+    async def route(self, message: BeamlimeMessage) -> None:
+        # Synchronous handlers
+        if handlers := self.handlers.get(type(message), []):
+            for handler in handlers:
+                await asyncio.sleep(0)  # Let others use the event loop.
+                with self._handler_wrapper(handler, message):
+                    self._check_result(handler(message))
+
+        # Asynchronous handlers
+        if awaitable_handlers := self.awaitable_handlers.get(type(message), []):
+            for handler in awaitable_handlers:
+                with self._handler_wrapper(handler, message):
+                    self._check_result(await handler(message))
```

Comment:
This means that you run the handlers in sequence, i.e., they cannot run in parallel. It would be better to spawn them all at the same time.

Reply:
That is true....!